pyleco/core/message.py

```diff
@@ -0,0 +1,242 @@
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+from json import JSONDecodeError
+from typing import Any, List, Optional, Self, Tuple
+
+
+from . import VERSION_B
+from .serialization import (create_header_frame, serialize_data, interpret_header, split_name,
+                            deserialize_data,
+                            )
+
+
+# Control transfer protocol
+class Message:
+    """A message in LECO protocol.
+
+    It consists of the following bytes frames, for which each an attribute exists:
+        - `version`
+        - `receiver`
+        - `sender`
+        - `header`
+        - 0 or more `payload` frames
+
+    If you do not specify a sender, the sending program shall add it itself.
+    The :attr:`data` attribute is the content of the first :attr:`payload` frame.
+    All attributes, except the official frames, are for convenience.
+    """
+
+    version: bytes = VERSION_B
+
+    def __init__(self, receiver: bytes | str, sender: bytes | str = b"",
+                 data: Optional[bytes | str | Any] = None,
+                 header: Optional[bytes] = None,
+                 conversation_id: Optional[bytes] = None,
+                 message_id: Optional[bytes] = None,
+                 message_type: Optional[bytes] = None,
+                 **kwargs) -> None:
+        self._setup_caches()
+        self.receiver = receiver if isinstance(receiver, bytes) else receiver.encode()
+        self.sender = sender if isinstance(sender, bytes) else sender.encode()
+        self.header = (create_header_frame(conversation_id=conversation_id, message_id=message_id,
+                                           message_type=message_type)
+                       if header is None else header)
+        if isinstance(data, (bytes)):
+            self.payload = [data]
+        if isinstance(data, str):
+            self.payload = [data.encode()]
+        else:
+            self._data = data
+
+    def _setup_caches(self) -> None:
```

Comment:
I use `None` to indicate, that the cache is empty and wanted to express it with that Optional.

So here `None | bytes` would be better?

Reply:
> I use None to indicate, that the cache is empty and wanted to express it with that Optional.

Why not take an empty object of the target type to indicate emptiness? This might be the typed-lang programmer in me, but this seems like a simpler design -- you don't switch types, the intent is still expressed, there is no need for `Optional`.

I'm not sure if the whole overhead of keeping track of several objects, checking emptiness, processing logic with the data attribute (when to process, emptying the correct one,...) is worth it to cache a couple of string operations -- compared to the network/cross-process access we are doing in here, I suspect those processing times will be quite negligible. Might make sense to measure first, to not [prematurely optimize](https://wiki.c2.com/?PrematureOptimization) and take the burden of a lot of logic for marginal benefit.

Reply:
I'll drop all the cached stuff. Only the Coordinator would profit and the coordinator could cache them itself.